lib/lotus/model/migrator.rb

```diff
@@ -0,0 +1,544 @@
+    # keep track of current migration version. If no migration is
+    # stored in the database, the version is considered to be 0.
+    # if no target version is specified, the database is migrated
+    # to the latest version available in migration directory.
+    #
+    # The implementation borrows much code from Sequel::Migrator.
+    #
+    # @example Migrate all migrations in db/migrations to latest
+    #   require 'lotus/model'
+    #   require 'lotus/model/migrator'
+    #
+    #   Lotus::Model.configure do
+    #     adapter type: :sql, uri: 'sqlite3://localhost/database'
+    #   end
+    #
+    #   Lotus::Model::Migrator.new.run
+    #
+    # @since x.x.x
+    # @see Sequel::Migration
+    class Migrator
+      # The framework configuration
+      #
+      # @since x.x.x
+      # @api private
+      include Utils::ClassAttribute
+      class_attribute :configuration
+
+      # It's raised when migration does not support the adapter
+      #
+      # @since x.x.x
+      class UnsupportedAdapterError < ::StandardError
+        def initialize(adapter_type)
+          super("Adapter #{adapter_type} is not supported")
+        end
+      end
+
+      # It's raised when adapter is not configured
+      #
+      # @since x.x.x
+      class MissingAdapterConfigurationError < ::StandardError
+        def initialize
+          super('Please configure your adapter. See Lotus::Model.configure for more details')
+        end
+      end
+
+      # It's raised when there are missing migration files
+      #
+      # @since x.x.x
+      class MissingMigrationsError < ::StandardError
+        def initialize(missing_migration_files)
+          super("Applied migration files not in file system: #{missing_migration_files.map { |f| f.path }.join(', ')}")
+        end
+      end
+
+      # Error for duplicated versions in migration files
+      # It's raised when there are duplication in versionings
+      #
+      # @since x.x.x
+      #
+      # @see Lotus::Model::Migrator#run
+      class DuplicatedVersionsError < ::StandardError
+      end
+
+      # The list of supported adapters
+      #
+      # @since x.x.x
+      SUPPORTED_ADAPTERS = [:sql].freeze
+
+      # The default column for schema migrations table
+      #
+      # @since x.x.x
+      SCHEMA_COLUMN = :version
+
+      # The default table for scheme migrations
+      #
+      # @since x.x.x
+      SCHEMA_TABLE = :schema_migrations
+
+      def initialize(opts={})
+        _copy_framework_configuration
+        _check_if_adapter_is_configured
+        _check_if_adapter_support_migration
+
+        @adapter = _get_adapter
+        @db = adapter.instance_variable_get(:@connection)
+        @logger = _get_logger
+        @directory = _get_migration_directory
+        @allow_missing_migration_files = opts[:allow_missing_migration_files]
+        @files = _get_migration_files
+        @column = SCHEMA_COLUMN
+        schema, table = @db.send(:schema_and_table, SCHEMA_TABLE)
```

Comment:
This denotes intimate knowledge of `Sequel`. If `@db` isn't our object, we should consider to introduce a `Connection` concept in `Model`. This looks dangerous in the long term.